tools/stats/upload_test_stats.py

```diff
@@ -0,0 +1,175 @@
+GITHUB_TOKEN = os.environ["GITHUB_TOKEN"]
+REQUEST_HEADERS = {
+    "Accept": "application/vnd.github.v3+json",
+    "Authorization": "token " + GITHUB_TOKEN,
+}
+S3_RESOURCE = boto3.resource("s3")
+
+
+def parse_xml_report(report: Path, workflow_id: int):
+    """Convert a test report xml file into a JSON-serializable list of test cases."""
+    # Retrieve the job id from the report path. In our GHA workflows, we append
+    # the job id to the end of the report name, so `report` looks like:
+    #     unzipped-test-reports-foo_5596745227/test/test-reports/foo/TEST-foo.xml
+    # and we want to get `5596745227` out of it.
+    job_id = int(report.parts[0].rpartition("_")[2])
+
+    print(f"Parsing test report: {report}, job id: {job_id}")
+    root = ET.parse(
+        report,
+        ET.XMLParser(target=ET.TreeBuilder(insert_comments=True)),  # type: ignore [call-arg]
+    )
+
+    test_cases = []
+    for test_case in root.findall("testcase"):
+        case = process_xml_element(test_case)
+        case["workflow_id"] = workflow_id
+        case["job_id"] = job_id
+        test_cases.append(case)
+
+    return test_cases
+
+
+def process_xml_element(element):
+    """Convert a test suite element into a JSON-serializable dict."""
+    ret = {}
+
+    # Convert attributes directly into dict elements.
+    # e.g.
+    #     <testcase name="test_foo" classname="test_bar"></testcase>
+    # becomes:
+    #     {"name": "test_foo", "classname": "test_bar"}
+    ret.update(element.attrib)
+
+    # By default, all attributes are strings. Apply a few special conversions
+    # here for well-known attributes so that they are the right type in Rockset.
+    if line := ret.get("line"):
+        ret["line"] = int(line)
+    if time := ret.get("time"):
+        ret["time"] = float(time)
+    if timestamp := ret.get("timestamp"):
+        # Timestamps reported are not valid ISO8601 because they have no timezone. Add one.
+        # This assumes that
+        ret["timestamp"] = (
+            datetime.datetime.fromisoformat(timestamp).astimezone().isoformat()
+        )
+
+    # Convert inner and outer text into special dict elements.
+    # e.g.
+    #     <testcase>my_inner_text</testcase> my_tail
+    # becomes:
+    #     {"text": "my_inner_text", "tail": " my_tail"}
+    if element.text and element.text.strip():
+        ret["text"] = element.text
+    if element.tail and element.tail.strip():
+        ret["tail"] = element.tail
+
+    # Convert child elements recursively, placing them at a key:
+    # e.g.
+    #     <testcase>
+    #       <foo>hello</foo>
+    #     </testcase>
+    # becomes
+    #    {"foo": {"text": "hello"}}
+    for child in element:
+        # Special handling for comments.
+        if child.tag is ET.Comment:
+            ret["comment"] = child.text
+        else:
+            ret[child.tag] = process_xml_element(child)
+    return ret
+
+
+def get_artifact_urls(workflow_run_id: int) -> Dict[Path, str]:
```

Comment:
Ah, is this written because artifact transfer is not a thing for workflow_runs? That would make sense.